python/hsfs/core/tfdata_engine.py

```diff
@@ -141,37 +150,123 @@ def _de_serialize(serialized_example):
+                else:
+                    x = example[_feature_name]
+                return x, y
+            # Otherwise we need to have features in the same type, thus tf.float32
+            else:
+                x = []
+                for _feature_name in self._feature_names:
+                    x.append(_convert2float32(example[_feature_name]))
+                x = tf.stack(x)
             return x, y
 
         dataset = dataset.map(
             lambda value: _de_serialize(value),
             num_parallel_calls=tf.data.experimental.AUTOTUNE,
         )
 
-        if optimize:
+        if process:
+            dataset = dataset.map(
+                lambda value: _process_example(value),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
             dataset = _optimize_dataset(
                 dataset, batch_size, num_epochs, self._is_training
             )
 
         return dataset
 
+    def tf_csv_dataset(
+        self,
+        batch_size=None,
+        num_epochs=None,
+        one_hot_encode_labels=False,
+        num_classes=None,
+        process=False,
+    ):
+        """
+        Reads and returns tf.data.TFRecordDataset object ready to feed tf keras models
+
+        :param batch_size: size of batch, defaults to None
+        :type batch_size: int, optional
+        :param num_epochs: number of epochs to train, defaults to None
+        :type num_epochs: int, optional
+        :param one_hot_encode_labels: if set true then one hot encode lables, defaults to False
+        :type one_hot_encode_labels: boolean, optional
+        :param num_classes: if above true then provide number of target classes, defaults to None
+        :type num_classes: int, optional
+        :param process: if set true  api will optimise tf data read operation, and return feature vector for model
+        with single input, defaults to False
+        :type  process: int, optional
+        :return: tf dataset
+        :rtype: tf.data.TFRecordDataset
+        """
+
+        if self._training_dataset_format != "csv":
+            raise ValueError(
+                "tf_csv_dataset function works only for training datasets that have csv format"
+            )
+
+        if process and (batch_size is None or num_epochs is None):
+            raise ValueError(
+                "if process is set to True you also need to provide batch_size and num_epochs"
+            )
+
+        if one_hot_encode_labels and (num_classes is None or num_classes <= 1):
+            raise ValueError(
+                "if one_hot_encode_labels is set to True you also need to provide num_classes > 1"
+            )
+
+        csv_header = [feat.name for feat in self._training_dataset_schema]
+        record_defaults = [
+            _convert2tfdtype(feat.type) for feat in self._training_dataset_schema
+        ]
+
+        csv_dataset = tf.data.experimental.CsvDataset(
```

Comment:
we should try to use as little as possible from the experimental modules. Can we use the higherlevel tf.data.Dataset?

Reply:
I thinkokay how I understood, also `tf.data.Dataset` is in memory and `CsvDataset` is just another implementation of it.

Reply:
or actually the other way around, Dataset is higher level, CsvDataset gives you a bit more control. But essentially they do the same thing, just that Dataset is in the stable API module.

Reply:
I believe it only keeps the dataset in memory if you create the `Dataset` be it the regular one or `CsvDataset` from an in-memory object such as a pandas dataframe with the from-tensor-slices method.